### PR TITLE
Bonnie 402 backbone 404 page routing

### PR DESCRIPTION
**Pull request by mayerm94**
Add default route to page not found error page and also route to the page not found page when a page for a non existent resource is requested. Handlebars template and Thorax view are added for the error page.

### CODE REVIEW

File: app/assets/javascripts/router.js.coffee [hunk 1]
```diff
@@ -98,41 +101,49 @@
...
-    measure.get('upload_summaries').loadCollection(true)
-      .done( (uploadSummaries) =>
-        @navigationSetup "Measure Upload History - #{measure.get('cms_id')}", 'test-case-history'
-        # @collection = new Thorax.Collections.Patients
-        @mainView.setView new Thorax.Views.MeasureHistoryView model: measure, patients: measure.get('patients'), upload_summaries: uploadSummaries
-        @breadcrumb.viewMeasureHistory(measure) )
-      .fail( => @showError title: "Measure History Load Failed", summary: "Historic data failed to load due to a server error." )
+    if measure? 
+      @mainView.setView new Thorax.Views.LoadingView
+      # show loading view because this data is loaded async. Show breadcrumb now so people know where they are heading.
```
Comment (edeyoung) [inline app/assets/javascripts/router.js.coffee:116]: this comment got moved around. recommend moving it back above line 115

File: app/assets/javascripts/router.js.coffee [hunk 2]
```diff
@@ -98,41 +101,49 @@
...
+      @breadcrumb.viewMeasureHistory(measure)
+      measure.get('upload_summaries').loadCollection(true)
+        .done( (uploadSummaries) =>
+          @navigationSetup "Measure Upload History - #{measure.get('cms_id')}", 'test-case-history'
+          # @collection = new Thorax.Collections.Patients
+          @mainView.setView new Thorax.Views.MeasureHistoryView model: measure, patients: measure.get('patients'), upload_summaries: uploadSummaries
+          @breadcrumb.viewMeasureHistory(measure) )
+          .fail( => @showError title: "Measure History Load Failed", summary: "Historic data failed to load due to a server error." )
```
Comment (edeyoung) [inline app/assets/javascripts/router.js.coffee]: indentation looks like it got messed up in the copy/paste

File: app/assets/javascripts/router.js.coffee [hunk 3]
```diff
@@ -31,13 +31,7 @@
...
     'admin/users':                                      'renderUsers'
     'value_sets/edit':                                  'renderValueSetsBuilder'
-
-  # TODO: most of these functions have nil checks on the measure but then continue
-  # to perform logic on the measure without doing any further nil checks. If a nil
-  # check is necessary, there should be a nice way to error out of the function
-  # completely.
-  # these functions should all be reworked to error out appropriately if a measure
-  # is nil.
+    '*path':                                            'showPageNotFound'
```
Comment (edeyoung) [inline app/assets/javascripts/router.js.coffee:35]: consider adding a comment explaining that this catches any unspecified routes.

(nice addition, by the way)

File: app/assets/javascripts/templates/error_view.hbs [hunk 4]
```diff
@@ -0,0 +1,15 @@
+<div class="error-page">
```
Comment (edeyoung) [inline app/assets/javascripts/templates/error_view.hbs:1]: @thecristen does this page need any 508 tags added to it or is the included text sufficient?

File: spec/javascripts/views/patient_builder_spec.js.coffee [hunk 5]
```diff
@@ -13,6 +13,13 @@ describe 'PatientBuilderView', ->
     spyOn(@patientBuilder.model, 'materialize')
     spyOn(@patientBuilder.originalModel, 'calculateAndSave').and.returnValue(true)
     @$el = @patientBuilder.$el
+    
+  it 'should not open patient builder for non existent measure', ->
+    spyOn(bonnie,'showPageNotFound')
+    bonnie.renderPatientBuilder('non_existant_hqmf_set_id',@patient.id)
```
Comment (edeyoung) [inline spec/javascripts/views/patient_builder_spec.js.coffee]: include a space after the comma. same in other parts of the code.

**Review by edeyoung — CHANGES_REQUESTED** [review]
Overall looks really good. some minor changes requested.

File: app/assets/javascripts/templates/error_view.hbs (hunk 4 above)
Comment (thecristen) [reply app/assets/javascripts/templates/error_view.hbs:1]: @edeyoung The included text is sufficient! 👍🏼

**Review by thecristen — APPROVED** [review]

**Review by edeyoung — APPROVED** [review]

File: app/assets/javascripts/templates/error_view.hbs [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+<h1 class="text-primary">Page Not Found</h1>
+<p class="text-center text-info">
+  <strong>
+    <i class="fa fa-fw fa-3x fa-question-circle" aria-hidden="true"></i>
+    <img src="/assets/Rabbit-96-blue.png" alt = "">
```
Comment (holmesie) [inline app/assets/javascripts/templates/error_view.hbs:6]: This won't work on production deployments of Rails, see: http://guides.rubyonrails.org/asset_pipeline.html#what-is-fingerprinting-and-why-should-i-care-questionmark

**Pull request merged**